asreviewcontrib/semantic_clustering/semantic_clustering.py

```diff
@@ -75,24 +83,73 @@ def SemanticClustering(asreview_data_object):
     print("Running t-SNE...")
     tsne = t_sne(pca, n_iter=1000)
 
+    # calculate optimal number of clusters
+    print("Calculating optimal number of clusters...")
+    n_clusters = _calc_optimal_n_clusters(tsne)
+    print("Optimal number of clusters: ", n_clusters)
+
     # run k-means
     print("Running k-means...")
-    labels = run_KMeans(tsne, 10, 10)
+    labels = run_KMeans(tsne, n_clusters, 10)
 
     # visualize clusters
     print("Visualizing clusters...")
-    tsne_data = [tsne[:, 0], tsne[:, 1]]
-    visualize_clusters(tsne_data, labels)
+    _visualize_clusters(tsne, labels)
+
+    # create file for use in interactive dashboard
+    _create_file(data, tsne, labels)
+
+
+# Create functional dataframe and store to file for use in interactive
+def _create_file(data, coords, labels):
+    data['x'] = coords[:, 0]
+    data['y'] = coords[:, 1]
+    data['cluster_id'] = labels
+
+    if not os.path.exists("data"):
+        os.makedirs("data")
+
+    kmeans_df_path = os.path.join("data", "kmeans_df.csv")
+    data.to_csv(kmeans_df_path, index=None)
+
+
+# Optimal n clusters, very inefficient, to be corrected in a future PR
+def _calc_optimal_n_clusters(features):
 
+    sum_of_squared_distances = []
 
-def visualize_clusters(data, labels):
+    K = range(1, 25)
```

Comment:
why 25?

Reply:
Explained in other PR. Will be merged soon!